File: app/src/main/java/com/bresiu/codechallenge/presentation/viewmodel/ListViewModel.java
package com.bresiu.codechallenge.presentation.viewmodel;

import android.util.Log;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;
import com.bresiu.codechallenge.model.PostWithUserAddress;
import com.bresiu.codechallenge.presentation.uimodels.Result;
import com.bresiu.codechallenge.presentation.uimodels.ResultBundle;
import com.bresiu.codechallenge.repository.Repository;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.BehaviorSubject;
import java.util.List;
import javax.inject.Inject;

public class ListViewModel extends ViewModel {
	private final Repository repository;
	private final CompositeDisposable compositeDisposable;
	private BehaviorSubject<Long> subject;
	private MediatorLiveData<Result<List<PostWithUserAddress>>> mediatorLiveData;

	@Inject ListViewModel(Repository repository) {
		this.repository = repository;
		compositeDisposable = new CompositeDisposable();
		initLiveData();
		initListeners();
		fetchData();
	}

	private void initListeners() {
		subject = BehaviorSubject.create();
		compositeDisposable.add(subject.subscribeOn(Schedulers.io())
				.observeOn(Schedulers.io())
				.subscribe(repository::deletePostById));
	}

	private void initLiveData() {
		Log.d("BRS", "initLiveData");
		mediatorLiveData = new MediatorLiveData<>();
		mediatorLiveData.addSource(repository.getLiveData(), postWithUserAddresses -> {
			Log.d("BRS", "onChanged");
			if (postWithUserAddresses.isEmpty()) {
				Log.d("BRS", "skip empty update");
			} else {
				mediatorLiveData.setValue(Result.successResult(new ResultBundle<>(postWithUserAddresses)));
			}
		});
		mediatorLiveData.setValue(Result.loadingResult());
	}

	@Override protected void onCleared() {
		compositeDisposable.dispose();
		super.onCleared();
	}

	public LiveData<Result<List<PostWithUserAddress>>> getLiveData() {
		return mediatorLiveData;
	}

	private void fetchData() {
		compositeDisposable.add(repository.fetchData()
				.subscribe(repository::saveData,
						throwable -> mediatorLiveData.setValue(Result.errorResult(throwable))));
	}

	public void deletePostById(long postId) {
		subject.onNext(postId);
	}
}

File: app/src/main/java/com/bresiu/codechallenge/presentation/viewmodel/ListViewModel.kt
package com.bresiu.codechallenge.presentation.viewmodel

import androidx.lifecycle.*
import com.bresiu.codechallenge.model.PostWithUser
import com.bresiu.codechallenge.presentation.uimodels.NoResultsFoundException
import com.bresiu.codechallenge.presentation.uimodels.Result
import com.bresiu.codechallenge.presentation.uimodels.ResultBundle
import com.bresiu.codechallenge.repository.Repository
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import javax.inject.Inject

class ListViewModel @Inject internal constructor(private val repository: Repository) : ViewModel() {
  private val compositeDisposable: CompositeDisposable = CompositeDisposable()
  private var subject: BehaviorSubject<Long> = BehaviorSubject.create()
  private var mediatorLiveData: MediatorLiveData<Result<List<PostWithUser>>> = MediatorLiveData()
  private val searchInput: MutableLiveData<String> = MutableLiveData()
  private val sourceSwitcher = Transformations.switchMap(searchInput) {
    if (it.isNotEmpty()) {
      repository.searchDataForPhrase(formatPhrase(it))
    } else {
      repository.getAllData()
    }
  }

  private fun formatPhrase(phrase: String): String {
    return "%$phrase%"
  }

  val liveData: LiveData<Result<List<PostWithUser>>>?
    get() = mediatorLiveData

  init {
    initLiveData()
    initListeners()
    fetchData()
  }

  private fun initListeners() {
    compositeDisposable.add(subject.subscribeOn(Schedulers.io())
        .observeOn(Schedulers.io())
        .subscribe { repository.deletePostById(it) })
  }

  private fun initLiveData() {
    mediatorLiveData.addSource(sourceSwitcher) { postWithUserAddresses ->
      postDataIfNotEmpty(postWithUserAddresses)
    }
    onSearchViewCollapsed()
    mediatorLiveData.value = Result.loadingResult()
  }

  override fun onCleared() {
    compositeDisposable.dispose()
    super.onCleared()
  }

  private fun postDataIfNotEmpty(data: List<PostWithUser>) {
    if (data.isNotEmpty()) {
      mediatorLiveData.postValue(Result.successResult(ResultBundle(data)))
    } else if (!searchInput.value.isNullOrBlank()) {
      mediatorLiveData.postValue(Result.errorResult(NoResultsFoundException()))
    }
  }

  private fun fetchData() {
    compositeDisposable.add(repository.fetchData()
        .subscribe({
          repository.saveData(it)
        }, {
          mediatorLiveData.postValue(Result.errorResult(it))
        }))
  }

  fun deletePostById(postId: Long) {
    subject.onNext(postId)
  }

  fun onSearchViewCollapsed() {
    searchInput.postValue("")
  }

  fun searchForData(query: String) {
    searchInput.postValue(query)
  }
}
